hypothesis-python/tests/cover/test_lookup.py

```diff
@@ -540,6 +540,72 @@ def test_resolving_recursive_type():
     assert isinstance(st.builds(Tree).example(), Tree)
 
 
+class MyList:
+    def __init__(self, nxt: typing.Optional["MyList"] = None):
+        self.nxt = nxt
+
+    def __repr__(self):
+        return f"MyList({self.nxt})"
+
+
+def test_resolving_recursive_type_with_defaults():
+    assert isinstance(st.from_type(MyList).example(), MyList)
+
+
+class A:
+    def __init__(self, nxt: typing.Optional["B"]):
+        self.nxt = nxt
+
+    def __repr__(self):
+        return f"A({self.nxt})"
+
+
+class B:
+    def __init__(self, nxt: typing.Optional["A"]):
+        self.nxt = nxt
+
+    def __repr__(self):
+        return f"B({self.nxt})"
+
+
+def test_resolving_mutually_recursive_types():
+    nxt = st.from_type(A).example()
```

Comment:
Let's use `@given()` for these tests, so that we can be confident that it works for many diverse inputs, rather than just whichever happend to be generated.

(this is _probably_ fine, but getting the first `.example()` takes just as long and I've been bitten by merging PRs without finding one-in-a-hundred counterexamples before)

Reply:
Will do.

The with-defaulted-arg variants will only ever produce a single value, but that's an issue for another day.

Reply:
Hmm, maybe dropping a `st.deferred()` in when we hit the recursion would do the trick?  If it really is that simple I'd prefer to include it in this PR; otherwise let's open a follow-up issue to make sure we don't lose track.

Reply:
It looks like it does the trick, yes. As does `...`, which looks slightly better IMO in terms of not duplicating logic, so I'll put that in.

Btw: There will be some force-resetting of this branch to figure out that pypy business. Please ignore.